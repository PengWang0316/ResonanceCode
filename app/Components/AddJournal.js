import React, { Component } from "react";
import Util from "../apis/Util";
import Loading from "./Loading"
import JournalContent from "./JournalContent";
import ReadingSearchAndList from "./ReadingSearchAndList";
import { isLogin } from "../apis/LoginApi";
import DatabaseApi from "../apis/DatabaseApi";
import PropTypes from "prop-types";

class AddJournal extends Component {

  componentWillMount(){
    // Initial variables
    this.readings = this.props.journalData ? this.props.journalData.readingIds : {}; // keep which readings should be attached on. Format is like {readingId: null}
    this.contents = {}; // Keep content keys and content
    this.contentKeyIndex = 0; // Generate keys for different contents
    this.contentIndexs = {}; // Use to track the index of content component in the array. Delete function needs it. The format is {contentKey: index}
    this.jounalId = this.props.journalData ? this.props.journalData._id : null; // Keeping journal id for update
    this.userId = isLogin(document).userid;

    this.state={
      isWriting: false,
      journalDate: this.props.journalData ? Util.getDateString(this.props.journalData.date) : Util.getCurrentDateString(),
      isDateCorrect: true,
      isEmptyReading: this.props.journalData ? false : true,
      contentComponentArray: [], // keep content component
      addJournalContent: "overview",
      pingPongState: this.props.journalData ? this.props.journalData.ping_pong_state : "Neutral"
    };

    console.log("Journal before alter: ", this.props.journalData);
    // If journal data exsit, get content
    if(this.props.journalData){
      // this.oldContentKeys = []; // keep original content keys
      this.oldReadingIds = Object.keys(this.readings); // keep original reading ids

      // delete unnecessary properties from journal object
      let journal = this.props.journalData;
      this.journalUserId = journal.user_id;
      delete journal.date;
      delete journal.user_id;
      delete journal.ping_pong_state;
      delete journal._id;
      delete journal.readingIds;
      let keyRegExp = /-(\d+)$/; // The regular expression for subtract suffixs
      console.log("Journal after alter: ", this.props.journalData);
      Object.keys(journal).map((key)=>{
        // this.oldContentKeys.push(key); // keeping the exsit keys in an array for delete function
        // handleAddContentClick(addJournalContent, newContentName, newContentKey, contentKeyIndex, isShared)
        // Also have to save the largest key number and use it to countine
        // journalData property is like {'overview-0': 'test overview', 'overview-0-isShared': true}
        // if key is isShared do not do anything
        let matchResult = key.match(keyRegExp);
        if (matchResult){
          if (this.contentKeyIndex < matchResult[1]) this.contentKeyIndex = matchResult[1]; // keep the max index number and user can contiune add new content
          this.handleAddContentClick(journal[key], key.replace(keyRegExp, ""), key, matchResult[1], journal[`${key}-isShared`]);
          // console.log("isShared: ",journal[`${key}-isShared`]);
        }
      });
      this.contentKeyIndex++;
    }

  }

  handleChange(event, element){
    let newState={};
    newState[element]=event.target.value;
    if(element=="journalDate") newState.isDateCorrect = Util.matchDateFormat(newState[element]);
    this.setState(newState);
  }

  handleChangeCallback(contentKey, contentText){
    console.log(contentKey, contentText);
    // update content in state
    this.contents[contentKey]=contentText;
  }

  handleDeleteContentCallback(contentKey){
    // Remove content from state
    if(this.contents.hasOwnProperty(contentKey)){
      delete this.contents[contentKey];
      delete this.contents[`${contentKey}-isShared`];
    }
    // Remove content component
    delete this.contentIndexs[contentKey];
    this.setComponentToStateArray();
    // this.state.contentComponentArray.splice(this.contentIndexs[contentKey],1);
    // this.setState({contentComponentArray: this.state.contentComponentArray});
  }

  handleSharedBoxChangeCallback(contentKey, isShared){
    this.contents[`${contentKey}-isShared`] = isShared;
  }

  handleAddContentClick(addJournalContent, newContentName, newContentKey, contentKeyIndex, isShared){
    // put component in array in order to show
    newContentKey = newContentKey ? newContentKey : `${this.state.addJournalContent}-${this.contentKeyIndex}`;
    // put the component in the track object
    this.contentIndexs[newContentKey]=<JournalContent key={contentKeyIndex ? contentKeyIndex : this.contentKeyIndex++} newContent={addJournalContent ? addJournalContent : ""} newContentName={newContentName ? newContentName : this.state.addJournalContent} newContentKey={newContentKey} handleChangeCallback={(contentKey, contentText)=>{this.handleChangeCallback(contentKey, contentText);}} handleDeleteContentCallback={(contentKey)=>{this.handleDeleteContentCallback(contentKey);}} handleSharedBoxChangeCallback={(contentKey, isShared)=>{this.handleSharedBoxChangeCallback(contentKey, isShared)}} isShared={isShared ? true : false} />;
    this.setComponentToStateArray();
    // this.setState({contentComponentArray: this.state.contentComponentArray});

    // also have to put content title to state in order to track content and update
    this.contents[newContentKey] = addJournalContent ? addJournalContent : "";
    this.contents[`${newContentKey}-isShared`] = isShared ? true : false;
  }

  setComponentToStateArray(){
    // transfor object to array in order to display
    this.state.contentComponentArray=[];
    Object.keys(this.contentIndexs).map((element)=>{
      this.state.contentComponentArray.push(this.contentIndexs[element]);
    });
    this.setState({contentComponentArray: this.state.contentComponentArray});
  }

  handleSubmit(event, isUpdate){
    event.preventDefault();
    // console.log("submit: contents:",this.contents);
    console.log("journal date:",this.state.journalDate);
    // Assemble a journal object for save
    let readingIdArray = [];
    Object.keys(this.readings).map((key)=>{readingIdArray.push(key)});
    let journal = Object.assign({_id:this.jounalId, date: new Date(this.state.journalDate), ping_pong_state: this.state.pingPongState, user_id: this.userId, readingIds: readingIdArray}, this.contents);

    if(isUpdate){
      console.log("submit: journal:", journal);
      // console.log("oldContentKeys:", this.oldContentKeys);
      console.log("oldReadingKeys:", this.oldReadingIds);
      // Assemble two arrays for deletion contents and reading
      let deleteContents = [];
      let deleteReadingIds = [];
      // this.oldContentKeys.map((element)=>{if(!this.contents.hasOwnProperty(element)) deleteContents.push(element);});
      this.oldReadingIds.map((element)=>{if(!this.readings.hasOwnProperty(element)) deleteReadingIds.push(element);});
      // journal.deleteContents = deleteContents; //todo delete deleteContents and oldContentKeys
      journal.deleteReadingIds = deleteReadingIds;
      console.log("delete contents: ",deleteContents);
      console.log("delete reading ids : ",deleteReadingIds);
      DatabaseApi.updateJournal(journal).then((result)=>{
        this.props.history.push("/reading");
      });
    }else{
      DatabaseApi.createJournal(journal).then((result)=>{
        this.props.history.push("/reading");
      });
    }
    /**/
    this.setState({isWriting: true});
    console.log("submit: journal:", journal);
  }

  /* For ReadingSearchAndList callback */
  handleAttachReadingCallback(readingId){
    this.readings[readingId] = null;
    this.setState({isEmptyReading: false});
  }

  handleDetachAttachReadingCallback(readingId){
    delete this.readings[readingId];
    this.setState({isEmptyReading: Object.keys(this.readings).length===0});
  }

  handleCancel(){
    this.props.history.push("/reading");
  }

  handleDelete(){
    console.log("Delete journal!");
  }

  render(){
    return(
      <div className="addReadingDiv">
        {this.state.isWriting && <Loading text="Creating" />}

        <div className="titleDiv">{this.props.journalData ? "Update journal for readings" : "Add a new journal for readings"}</div>
        <form className="form-horizontal" onSubmit={(event) => {this.handleSubmit(event, this.props.journalData);}}>

          <div className="text-right bottom-btn-div">
            {(!this.props.journalData || this.userId == this.journalUserId) &&
              <button type="submit" className="btn btn-info loginButton" disabled={this.state.isWriting || !(this.state.journalDate.length>0) || !(this.state.isDateCorrect) || this.state.isEmptyReading}>{this.props.journalData ? "Update" : "Submit"}</button>
            }
            {(this.props.journalData && this.userId == this.journalUserId) &&
              <button onClick={()=>{this.handleDelete()}} type="button" className="btn btn-danger loginButton">Delete</button>
            }
            <button onClick={()=>{this.handleCancel()}} type="button" className="btn btn-nomal loginButton">Cancel</button>
          </div>

            <div className="form-group row form-div">
              <label htmlFor="journalDate" className="col-sm-1 col-form-label">Date</label>
              <div className="col-sm-4">

                <input className={this.state.isDateCorrect?"form-control":"form-control form-control-warning"} type="text" placeholder="mm/dd/yyyy" id="journalDate" value={this.state.journalDate} onChange={(event)=>{this.handleChange(event, "journalDate")}} />
                {!this.state.isDateCorrect && <span className="glyphicon glyphicon-warning-sign form-control-feedback form-control-warning-span"></span>}

              </div>
              <label htmlFor="pingPongState" className="col-sm-3 col-form-label">Ping Pong State</label>
              <div className="col-sm-4">
                <select value={this.state.pingPongState} id="pingPongState" className="form-control" onChange={(event)=>{this.handleChange(event,"pingPongState");}}>
                  <option value="Neutral">Neutral</option>
                  <option value="Reading/Call">Reading/Call</option>
                  <option value="Impact absorbing">Impact absorbing</option>
                  <option value="Impact complete">Impact complete</option>
                  <option value="Response">Response</option>
                  <option value="Outward moving">Outward moving</option>
                  <option value="Outward complete">Outward complete</option>
                </select>
              </div>
            </div>

            {/* New content goes here */}
            {this.state.contentComponentArray}

            {/* Add content button and drop list */}
            <div className="row addJournalContentDiv">

              <div onClick={()=>{this.handleAddContentClick();}} className="addJournalContentBtnDiv col-sm-6"><i className="fa fa-plus-square" /> Add one content for your journal</div>
              <div className="col-sm-6">
                <select  className="form-control" onChange={(event)=>{this.handleChange(event,"addJournalContent");}}>
                  <option value="overview">Overview</option>
                  <option value="internal_environment">Internal Environment</option>
                  <option value="relational_environment">Relational Environment</option>
                  <option value="physical_environment">Physical Environment</option>
                  <option value="creative_vector">Creative Vector</option>
                  <option value="synchronicities">Synchronicities</option>
                  <option value="dreams">Dreams</option>
                  <option value="other">Other</option>
                </select>

              </div>

            </div>


            {/*  Start reading search function */}
            <ReadingSearchAndList readings={this.props.journalData ? this.readings : null} attachReadingCallback={(readingId)=>{this.handleAttachReadingCallback(readingId);}} detachReadingCallback={(readingId)=>{this.handleDetachAttachReadingCallback(readingId);}} />

              <div className="text-right bottom-btn-div">
                {(!this.props.journalData || this.userId == this.journalUserId) &&
                  <button type="submit" className="btn btn-info loginButton" disabled={this.state.isWriting || !(this.state.journalDate.length>0) || !(this.state.isDateCorrect) || this.state.isEmptyReading}>{this.props.journalData ? "Update" : "Submit"}</button>
                }
                {(this.props.journalData && this.userId == this.journalUserId) &&
                  <button onClick={()=>{this.handleDelete()}} type="button" className="btn btn-danger loginButton">Delete</button>
                }
                <button onClick={()=>{this.handleCancel()}} type="button" className="btn btn-nomal loginButton">Cancel</button>
              </div>
          </form>

        </div>
    );
  }


}
AddJournal.propTypes={
  journalData: PropTypes.object,
  history: PropTypes.object
};
export default AddJournal;
