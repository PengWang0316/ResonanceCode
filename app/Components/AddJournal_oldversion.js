import React, { Component } from "react";
import QueryString from "query-string";
import PropTypes from "prop-types";
import Util from "../apis/Util";
// import api from "../apis/api";
import DatabaseApi from "../apis/DatabaseApi";
import Loading from "./Loading";
import LoginApi from "../apis/LoginApi";

class AddJournal_oldversion extends Component{

  componentWillMount(){

    if(this.props.location){
      let queryInfo = QueryString.parse(this.props.location.search);
      this.readingId = queryInfo.readingId;
      this.readingName = queryInfo.readingName;
      this.readingDate = Util.getDateString(queryInfo.readingDate);
      // console.log("AddJournal paper readingId, readingName, and readingDate:", `Id: ${this.readingId} Name: ${this.readingName} Date: ${this.readingDate}`);
    }
    this.userId=LoginApi.isLogin(document).userid;

    let journalData = this.props.journalData;
    if(journalData){
      this.journalId=journalData.journalId;
      // this.readingDate=journalData.readingDate;
      this.readingName=journalData.readingName;
    }
    // console.log("AddJournal page:",journalData);
    // setting state
    this.state={
      isWriting: false,
      journalDate: journalData?Util.getDateString(journalData.date):Util.getCurrentDateString(),
      pingPongState: journalData?journalData.pingPongState:"Neutral",
      overview: journalData?journalData.overview:"",
      internalEnvironment: journalData?journalData.internalEnvironment:"",
      relationalEnvironment: journalData?journalData.relationalEnvironment:"",
      physicalEnvironment: journalData?journalData.physicalEnvironment:"",
      creativeVector: journalData?journalData.creativeVector:"",
      synchronicities: journalData?journalData.synchronicities:"",
      dreams: journalData?journalData.dreams:"",
      journalId: journalData?journalData.journalId:"",
      userId: journalData?journalData.userId:"",
      readingId: journalData?journalData.readingId:""
    };
    // if(this.journalId) this.setState(this.state);
  }

  componentDidMount(){
    // console.log("did mount");
    if(this.journalId) this.setState(this.state);
  }

  handleChange(event, element){
    // console.log(event.target.value);
    let newState={};
    newState[element]=event.target.value;
    this.setState(newState);
    // console.log(newState);
  }

  handleSubmit(event){
    event.preventDefault();
    // console.log("**********submit************");

      this.setState({isWriting: true});
      // console.log("submit:",this.state);
      // assemble object for creating
      let journalObject = {
        user_id: this.userid,
        reading_id: this.readingId?this.readingId:this.state.readingId,
        date: new Date(this.state.journalDate),
        overview: this.state.overview,
        internal_environment: this.state.internalEnvironment,
        relational_environment: this.state.relationalEnvironment,
        physical_environment: this.state.physicalEnvironment,
        creative_vector: this.state.creativeVector,
        synchronicities: this.state.synchronicities,
        dreams: this.state.dreams,
        ping_pong_state:this.state.pingPongState
      };
      // console.log("journalId", this.journalId);
      if(this.journalId) DatabaseApi.updateJournal(journalObject, this.state.journalId).then((reault)=>{this.props.history.push("/reading")});
      else DatabaseApi.createJournal(journalObject).then((reault)=>{this.props.history.push("/reading")});

  }

  handleDelete(event){
    event.preventDefault();
    // console.log("**********delete************");
    DatabaseApi.deleteJournal(this.state.readingId, this.state.journalId).then((reault)=>{this.props.history.push("/reading")});
  }

  render(){
    let titleDiv;
    if (this.readingId) titleDiv = <div><span className="rcTitle">Add a new journal to your reading: </span>{this.readingName}  ({this.readingDate})</div>;
      else titleDiv = <div><span className="rcTitle">Update {this.readingName}</span></div>;
    return(
      <div className="addReadingDiv">
        {this.state.isWriting && <Loading text="Creating" />}

        {titleDiv}
        <form className="form-horizontal" onSubmit={(event) => {this.handleSubmit(event);}}>

          <div className="text-right bottom-btn-div">{(!this.readingId && this.state.userId==this.userId) && <button type="button" className="btn btn-danger btn-del" onClick={(event)=>{this.handleDelete(event);}}>Delete</button>}{(this.readingId || this.state.userId==this.userId) && <button type="submit" className="btn btn-info loginButton" disabled={this.state.isWriting || !(this.state.journalDate.length>0) || !(this.state.overview.length>0)}>{this.readingId?"Submit":"Update"}</button>}</div>

            <div className="form-group row form-div">
              <label htmlFor="journalDate" className="col-sm-1 col-form-label">Date</label>
              <div className="col-sm-4">
                <input className="form-control" type="text" value={this.state.journalDate} id="journalDate" onChange={(event)=>{this.handleChange(event,"journalDate");}} />
              </div>
              <label htmlFor="pingPongState" className="col-sm-3 col-form-label">Ping Pong State</label>
              <div className="col-sm-4">
                <select value={this.state.pingPongState} id="pingPongState" className="form-control" onChange={(event)=>{this.handleChange(event,"pingPongState");}}>
                  <option value="Neutral">Neutral</option>
                  <option value="Reading/Call">Reading/Call</option>
                  <option value="Impact absorbing">Impact absorbing</option>
                  <option value="Impact complete">Impact complete</option>
                  <option value="Response">Response</option>
                  <option value="Outward moving">Outward moving</option>
                  <option value="Outward complete">Outward complete</option>
                </select>
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="overview" className="col-form-label">Overview</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.overview} placeholder="Overview..." id="overview" onChange={(event)=>{this.handleChange(event,"overview");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="internalEnvironment" className="col-form-label">Internal Environment</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.internalEnvironment} placeholder="Internal Environment..." id="internalEnvironment" onChange={(event)=>{this.handleChange(event,"internalEnvironment");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="relationalEnvironment" className="col-form-label">Relational Environment</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.relationalEnvironment} placeholder="Relational Environment..." id="relationalEnvironment" onChange={(event)=>{this.handleChange(event,"relationalEnvironment");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="physicalEnvironment" className="col-form-label">Physical Environment</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.physicalEnvironment} placeholder="Physical Environment..." id="physicalEnvironment" onChange={(event)=>{this.handleChange(event,"physicalEnvironment");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="creativeVector" className="col-form-label">Creative Vector</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.creativeVector} placeholder="Creative Vector..." id="creativeVector" onChange={(event)=>{this.handleChange(event,"creativeVector");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="synchronicities" className="col-form-label">Synchronicities</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.synchronicities} placeholder="Synchronicities..." id="synchronicities" onChange={(event)=>{this.handleChange(event,"synchronicities");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="dreams" className="col-form-label">Dreams</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.dreams} placeholder="Dreams..." id="dreams" onChange={(event)=>{this.handleChange(event,"dreams");}} />
              </div>
            </div>

          <div className="text-right bottom-btn-div">{(!this.readingId && this.state.userId==this.userId) && <button className="btn btn-danger btn-del" onClick={(event)=>{this.handleDelete(event);}}>Delete</button>}{(this.readingId || this.state.userId==this.userId) && <button type="submit" className="btn btn-info loginButton" disabled={this.state.isWriting || !(this.state.journalDate.length>0) || !(this.state.overview.length>0)}>{this.readingId?"Submit":"Update"}</button>}</div>

        </form>

      </div>
    );
  }

}
AddJournal_oldversion.propTypes={
  journalData: PropTypes.object
};
export default AddJournal_oldversion;
