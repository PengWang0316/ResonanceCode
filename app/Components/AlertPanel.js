import React from 'react';

import styles from '../styles/AlertPanel.module.css';

const AlertPanel = ({
  id = '', type = 'success', children = '', style = {}
}) => (
  <div id={id} className={`alert alert-${type} mt-3 ${styles.alertPanel}`} style={style} role="alert">
    {children}
  </div>
);
export default AlertPanel;
