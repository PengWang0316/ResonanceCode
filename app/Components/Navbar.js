import React from 'react';
import { connect } from 'react-redux';
import { NavLink } from 'react-router-dom';
import PropTypes from 'prop-types';

import logoImage from '../imgs/logo.png';
import facebookLogin from '../imgs/facebookLogin.png';
import googleLogin from '../imgs/googleLogin.png';
import signupButton from '../imgs/signup.png';
import { logout } from '../actions/UserActions';
import { API_FACEBOOK_LOGIN, API_GOOGLE_LOGIN } from '../actions/ApiUrls';

export const Navbar = ({ user, userLogout }) => (
  <nav className="navbar navbar-expand-md navbar-light bg-light">
    <div className="container">
      <NavLink activeClassName="navbar-brand" to="/"><img src={logoImage} alt="KairoScope" title="KairoScope" /></NavLink>

      <button className="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarNavDropdown" aria-controls="navbarNavDropdown" aria-expanded="false" aria-label="Toggle navigation">
        <span className="navbar-toggler-icon" />
      </button>


      <div className="collapse navbar-collapse" id="navbarNavDropdown">
        <div className="container-fluid d-flex justify-content-end">
          <ul className="navbar-nav">
            <li className="nav-item dropdown">
              <a className="nav-link dropdown-toggle" href="#" id="readingsDropdown" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">READINGS </a>
              <div className="dropdown-menu" aria-labelledby="readingsDropdown">
                <NavLink className="nav-link dropdown-item" activeClassName="active" to="/reading">YOUR READINGS</NavLink>
                <NavLink className="nav-link dropdown-item" activeClassName="active" to="/sharedReadings">SHARED READINGS</NavLink>
              </div>
            </li>
            <li className="nav-item dropdown">
              <a className="nav-link dropdown-toggle" href="#" id="navbarDropdownMenuLink" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">SEARCH </a>
              <div className="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
                <NavLink className="nav-link dropdown-item" activeClassName="active" to="/readingSearch">SEARCH READINGS</NavLink>
                <NavLink className="nav-link dropdown-item" activeClassName="active" to="/hexagramsSearch">SEARCH HEXAGRAMS</NavLink>
                <NavLink className="nav-link dropdown-item" activeClassName="active" to={{ pathname: '/allJournal' }}>ATTACHED JOURNALS</NavLink>
                <NavLink className="nav-link dropdown-item" activeClassName="active" to={{ pathname: '/unattachedJournals', search: '?readingName=Unattached Journals' }}>UNATTACHED JOURNALS</NavLink>
              </div>
            </li>

            <li className="nav-item"><NavLink className="nav-link" activeClassName="active" to="/hexagramList">HEXAGRAMS</NavLink></li>

            <li className="nav-item"><NavLink className="nav-link" activeClassName="active" to="/help">HELP</NavLink></li>

            {!user.isAuth &&
              <li className="nav-item dropdown">
                <a className="nav-link dropdown-toggle" href="#" id="socialLoginDropdown" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">SOCIAL LOGIN </a>
                <div className="dropdown-menu" aria-labelledby="socialLoginDropdown">
                  <a className="nav-link dropdown-item" href={API_FACEBOOK_LOGIN}><img alt="Facebook login button" src={facebookLogin} /></a>
                  <a className="nav-link dropdown-item" href={API_GOOGLE_LOGIN}><img alt="Google login button" src={googleLogin} /></a>
                  <NavLink className="nav-link dropdown-item" activeClassName="active" to="/signup"><img alt="Signup button" src={signupButton} /></NavLink>
                </div>
              </li>
            }

            {user.isAuth &&
              <li className="nav-item dropdown">
                <a className="nav-link dropdown-toggle" href="#" id="signoutDropdown" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">{user.photo && <img alt="User Avatar" className="avatar-photo" src={user.photo} />}
                  {!user.photo && <i className="fa fa-user-circle" aria-hidden="true" />} {user.displayName}
                </a>
                <div className="dropdown-menu" aria-labelledby="signoutDropdown">
                  <NavLink className="nav-link dropdown-item" activeClassName="active" to="/settings">SETTINGS</NavLink>
                  <NavLink className="nav-link" activeClassName="active" to="/" onClick={userLogout}>LOGOUT</NavLink>
                </div>
              </li>
            }

          </ul>
        </div>

      </div>

    </div>
  </nav>

);
Navbar.propTypes = {
  user: PropTypes.object.isRequired,
  userLogout: PropTypes.func.isRequired
};
/* istanbul ignore next */
const mapStateToProps = state => ({
  user: state.user
});
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  userLogout: _ => dispatch(logout())
});
export default connect(mapStateToProps, mapDispatchToProps, null, { pure: false })(Navbar);
