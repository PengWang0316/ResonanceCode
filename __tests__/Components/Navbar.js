import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { Navbar } from '../../app/Components/Navbar';

jest.mock('react-router-dom', () => ({ NavLink: 'NavLink' }));

describe('Navbar test', () => {
  const defaultProps = {
    user: { isAuth: true, photo: 'photoSrc', displayName: 'displayName' },
    userLogout: jest.fn()
  };
  const getShallowComponent = (props = defaultProps) => shallow(<Navbar {...props} />);

  test('Navbar login snapshot', () => expect(renderer.create(<Navbar {...defaultProps} />).toJSON()).toMatchSnapshot());

  test('Navbar not login snapshot', () => expect(renderer.create(<Navbar {...{ ...defaultProps, user: { isAuth: false } }} />).toJSON()).toMatchSnapshot());

  test('logout', () => {
    const component = getShallowComponent({
      ...defaultProps,
      user: { ...defaultProps.user, photo: null }
    });
    component.find({ to: '/' }).at(1).simulate('click');
    expect(defaultProps.userLogout).toHaveBeenCalledTimes(1);
  });
});
