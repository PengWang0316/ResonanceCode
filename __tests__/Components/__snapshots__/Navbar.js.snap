// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar test Navbar login snapshot 1`] = `
<nav
  className="navbar navbar-expand-md navbar-light bg-light"
>
  <div
    className="container"
  >
    <NavLink
      activeClassName="navbar-brand"
      to="/"
    >
      <img
        alt="KairoScope"
        src={
          Object {
            "process": [Function],
          }
        }
        title="KairoScope"
      />
    </NavLink>
    <button
      aria-controls="navbarNavDropdown"
      aria-expanded="false"
      aria-label="Toggle navigation"
      className="navbar-toggler"
      data-target="#navbarNavDropdown"
      data-toggle="collapse"
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="collapse navbar-collapse"
      id="navbarNavDropdown"
    >
      <div
        className="container-fluid d-flex justify-content-end"
      >
        <ul
          className="navbar-nav"
        >
          <li
            className="nav-item dropdown"
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              className="nav-link dropdown-toggle"
              data-toggle="dropdown"
              href="#"
              id="readingsDropdown"
            >
              READINGS 
            </a>
            <div
              aria-labelledby="readingsDropdown"
              className="dropdown-menu"
            >
              <NavLink
                activeClassName="active"
                className="nav-link dropdown-item"
                to="/reading"
              >
                YOUR READINGS
              </NavLink>
              <NavLink
                activeClassName="active"
                className="nav-link dropdown-item"
                to="/sharedReadings"
              >
                SHARED READINGS
              </NavLink>
            </div>
          </li>
          <li
            className="nav-item dropdown"
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              className="nav-link dropdown-toggle"
              data-toggle="dropdown"
              href="#"
              id="navbarDropdownMenuLink"
            >
              SEARCH 
            </a>
            <div
              aria-labelledby="navbarDropdownMenuLink"
              className="dropdown-menu"
            >
              <NavLink
                activeClassName="active"
                className="nav-link dropdown-item"
                to="/readingSearch"
              >
                SEARCH READINGS
              </NavLink>
              <NavLink
                activeClassName="active"
                className="nav-link dropdown-item"
                to="/hexagramsSearch"
              >
                SEARCH HEXAGRAMS
              </NavLink>
              <NavLink
                activeClassName="active"
                className="nav-link dropdown-item"
                to={
                  Object {
                    "pathname": "/allJournal",
                  }
                }
              >
                ATTACHED JOURNALS
              </NavLink>
              <NavLink
                activeClassName="active"
                className="nav-link dropdown-item"
                to={
                  Object {
                    "pathname": "/unattachedJournals",
                    "search": "?readingName=Unattached Journals",
                  }
                }
              >
                UNATTACHED JOURNALS
              </NavLink>
            </div>
          </li>
          <li
            className="nav-item"
          >
            <NavLink
              activeClassName="active"
              className="nav-link"
              to="/hexagramList"
            >
              HEXAGRAMS
            </NavLink>
          </li>
          <li
            className="nav-item dropdown"
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              className="nav-link dropdown-toggle"
              data-toggle="dropdown"
              href="#"
              id="signoutDropdown"
            >
              <img
                alt="User Avatar"
                className="avatar-photo"
                src="photoSrc"
              />
               
              displayName
            </a>
            <div
              aria-labelledby="signoutDropdown"
              className="dropdown-menu"
            >
              <NavLink
                activeClassName="active"
                className="nav-link dropdown-item"
                to="/settings"
              >
                SETTINGS
              </NavLink>
              <NavLink
                activeClassName="active"
                className="nav-link"
                onClick={[MockFunction]}
                to="/"
              >
                LOGOUT
              </NavLink>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;

exports[`Navbar test Navbar not login snapshot 1`] = `
<nav
  className="navbar navbar-expand-md navbar-light bg-light"
>
  <div
    className="container"
  >
    <NavLink
      activeClassName="navbar-brand"
      to="/"
    >
      <img
        alt="KairoScope"
        src={
          Object {
            "process": [Function],
          }
        }
        title="KairoScope"
      />
    </NavLink>
    <button
      aria-controls="navbarNavDropdown"
      aria-expanded="false"
      aria-label="Toggle navigation"
      className="navbar-toggler"
      data-target="#navbarNavDropdown"
      data-toggle="collapse"
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="collapse navbar-collapse"
      id="navbarNavDropdown"
    >
      <div
        className="container-fluid d-flex justify-content-end"
      >
        <ul
          className="navbar-nav"
        >
          <li
            className="nav-item dropdown"
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              className="nav-link dropdown-toggle"
              data-toggle="dropdown"
              href="#"
              id="readingsDropdown"
            >
              READINGS 
            </a>
            <div
              aria-labelledby="readingsDropdown"
              className="dropdown-menu"
            >
              <NavLink
                activeClassName="active"
                className="nav-link dropdown-item"
                to="/reading"
              >
                YOUR READINGS
              </NavLink>
              <NavLink
                activeClassName="active"
                className="nav-link dropdown-item"
                to="/sharedReadings"
              >
                SHARED READINGS
              </NavLink>
            </div>
          </li>
          <li
            className="nav-item dropdown"
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              className="nav-link dropdown-toggle"
              data-toggle="dropdown"
              href="#"
              id="navbarDropdownMenuLink"
            >
              SEARCH 
            </a>
            <div
              aria-labelledby="navbarDropdownMenuLink"
              className="dropdown-menu"
            >
              <NavLink
                activeClassName="active"
                className="nav-link dropdown-item"
                to="/readingSearch"
              >
                SEARCH READINGS
              </NavLink>
              <NavLink
                activeClassName="active"
                className="nav-link dropdown-item"
                to="/hexagramsSearch"
              >
                SEARCH HEXAGRAMS
              </NavLink>
              <NavLink
                activeClassName="active"
                className="nav-link dropdown-item"
                to={
                  Object {
                    "pathname": "/allJournal",
                  }
                }
              >
                ATTACHED JOURNALS
              </NavLink>
              <NavLink
                activeClassName="active"
                className="nav-link dropdown-item"
                to={
                  Object {
                    "pathname": "/unattachedJournals",
                    "search": "?readingName=Unattached Journals",
                  }
                }
              >
                UNATTACHED JOURNALS
              </NavLink>
            </div>
          </li>
          <li
            className="nav-item"
          >
            <NavLink
              activeClassName="active"
              className="nav-link"
              to="/hexagramList"
            >
              HEXAGRAMS
            </NavLink>
          </li>
          <li
            className="nav-item dropdown"
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              className="nav-link dropdown-toggle"
              data-toggle="dropdown"
              href="#"
              id="socialLoginDropdown"
            >
              SOCIAL LOGIN 
            </a>
            <div
              aria-labelledby="socialLoginDropdown"
              className="dropdown-menu"
            >
              <a
                className="nav-link dropdown-item"
                href="https://kairoscope.resonancepath.com:8080/api/v1/auth/facebookLogin"
              >
                <img
                  alt="Facebook login button"
                  src={
                    Object {
                      "process": [Function],
                    }
                  }
                />
              </a>
              <a
                className="nav-link dropdown-item"
                href="https://kairoscope.resonancepath.com:8080/api/v1/auth/googleLogin"
              >
                <img
                  alt="Google login button"
                  src={
                    Object {
                      "process": [Function],
                    }
                  }
                />
              </a>
              <NavLink
                activeClassName="active"
                className="nav-link dropdown-item"
                to="/signup"
              >
                <img
                  alt="Signup button"
                  src={
                    Object {
                      "process": [Function],
                    }
                  }
                />
              </NavLink>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;
